ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -437,7 +442,7 @@ func (r *RayJobReconciler) createNewK8sJob(ctx context.Context, rayJobInstance *
 			// is attempted 3 times at the maximum, but still mitigates the case of unrecoverable
 			// application-level errors, where the maximum number of retries is reached, and the job
 			// completion time increases with no benefits, but wasted resource cycles.
-			BackoffLimit: pointer.Int32(2),
+			BackoffLimit: backoffLimit,
```

Comment:
Personally I think this backoff limit should be part of the RayJob spec, what do you think? 

Some previous discussion on backoff limits here https://github.com/ray-project/kuberay/issues/1902#issuecomment-1925784679

Reply:
cc @kevin85421

Reply:
I agree this provides the most flexibility. Do we have use case where people want to set different limits for different RayJob?

Reply:
> Personally I think this backoff limit should be part of the RayJob spec, what do you think?

Currently, the purpose of the retry mechanism at this moment is to handle network instability in the Kubernetes cluster. For example, if the submitter temporarily cannot connect to the Ray head due to network issues, the retry mechanism may alleviate the problem by resubmitting the request. If the Ray head has a running or terminated Ray job, the retry will fail due to a conflict with the submission ID.

Therefore, it relates to the Kubernetes cluster that the RayJob CRs running in rather than to different Ray applications. That's why this PR sets the value for all RayJob CRs in the same Kubernetes cluster.

> https://github.com/ray-project/kuberay/issues/1902#issuecomment-1925784679
checks if there is job running with a submission id
> * if yes, starts tailing the logs => Ray currently doesn't support this.

Currently, Ray does not support tailing logs for an existing submission ID using `ray job submit`. If Ray were to support this feature, the retry mechanism could handle issues that arise after the Ray job starts running. In this case, making the `backoffLimit` a part of the RayJob spec makes sense to me because different Ray applications may require varying lengths of time to finish, but currently, Ray doesn't support that.

> Do we have use case where people want to set different limits for different RayJob?

Currently, this is not possible due to the ray job submit limitations I mentioned above.

My current thought is:
* Step 1: Support cluster-wide `backoffLimit` (i.e., this PR).
* Step 2: Enable `ray job submit` to tail logs if the submission ID exists.
* Step 3: Support `backoffLimit` as part of the RayJob spec, which can overwrite cluster-wide configuration.

What do you think?

Reply:
If (3) is the eventual goal, should we just do step 3 and then step 2 and skip step 1? Once we have the per job configuration, do we still want the cluster-wide configuration?

Reply:
I am fine with all options: (1) Step 1 -> Step 2 -> Step 3, (2) Step 2 -> Step 3, or (3) Step 3 -> Step 2.

Reply:
Let me then change this PR to do step 3 directly @andrewsykim ?

Reply:
>  If the Ray head has a running or terminated Ray job, the retry will fail due to a conflict with the submission ID.

I forgot about this problem. If we allow users to configure the back off limit, we need to also let them configure whether a new submissionID is generated for each retry. That probably needs a new field `submissionRetryPolicy` or something like that.

Reply:
> Let me then change this PR to do step 3 directly

LGTM since the retry policy can be done in a separate PR from this change anyways

Reply:
> If we allow users to configure the back off limit, we need to also let them configure whether a new submissionID is generated for each retry.

If the retry is caused by application-level logic, we should delete the RayCluster and create a new one for the retry.

Reply:
ack -- let's continue discussing more advanced RayJob retry policy in https://github.com/ray-project/kuberay/issues/1902